### PR TITLE
setup.sh: docker_container first, then fall back to docker_image

### PR DESCRIPTION
**Pull request by NorseGaud**
# Description

Here is an original discussion about this: https://github.com/docker-mailserver/docker-mailserver/pull/1874#issuecomment-809781531

Then there is https://github.com/docker-mailserver/docker-mailserver/pull/2080#discussion_r667359915: As seen in the comment thread, a user experienced confusion and didn't realize `USE_CONTAINER=true` or `-c` was necessary. With my current knowledge I can't see any reason why we wouldn't use the container by default and if it exists, fall back to using the default image/tag when it doesn't.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Improvement (non-breaking change that does improve existing functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (README.md or the documentation under `docs/`)
- [ ] If necessary I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

**Review by georglauterbach — APPROVED** [review]
LGTM 👍

File: setup.sh
Comment (NorseGaud) [reply setup.sh:224]: Sorry, can you clarify a bit what you're asking? I don't understand

Comment (NorseGaud) [reply setup.sh]: You're right... Unsure how I missed that

**Comment by NorseGaud** [comment]
Very interesting -- a test failed, but I can't reproduce it locally. Seems like there is some sort of flakeyness

**Comment by georglauterbach** [comment]
I've restarted the test suite. Let's see.

**Review by georglauterbach — DISMISSED** [review]

**Comment by NorseGaud** [comment]
@wernerfred any way that I can get access to re-run or cancel actions stuff? I think I already asked that, but I wanted to double check.

**Comment by wernerfred** [comment]
> @wernerfred any way that I can get access to re-run or cancel actions stuff? I think I already asked that, but I wanted to double check.

if you did so, i missed it!

imho nothing speaks against "promoting" you to be a maintainer which would include that permission. Another way would be to create a new team with write permission to the repository (kind of between triage and maintainer from a permission perspective)  if you prefer not to activly contribute as a maintainer but are slowed down in your contributions through the current permission level.

cc @georglauterbach

**Comment by NorseGaud** [comment]
> > @wernerfred any way that I can get access to re-run or cancel actions stuff? I think I already asked that, but I wanted to double check.
> 
> if you did so, i missed it!
> 
> imho nothing speaks against "promoting" you to be a maintainer which would include that permission. Another way would be to create a new team with write permission to the repository (kind of between triage and maintainer from a permission perspective) if you prefer not to activly contribute as a maintainer but are slowed down in your contributions through the current permission level.
> 
> image
> 
> cc @georglauterbach

Maintainer works for me. I don't see myself ever not using this repo to host my personal mail server :)

I'll probably be doing some stress testing on the test suite too since the more I use it the more I see random failures that don't reoccur, so this is perfect time to be talking about this.

**Comment by georglauterbach** [comment]
@wernerfred I'm fine with @NorseGaud becoming a maintainer :)

**Comment by wernerfred** [comment]
All set 🚀

**Comment by NorseGaud** [comment]
> All set 🚀

Gracias, bros!

**Comment by NorseGaud** [comment]
Tests are definitely flakey, and it's usually the same ones. I'll keep stress testing them and see if I can figure out why it's failing.

**Comment by NorseGaud** [comment]
Tested just the set of tests around what is failing and can't get it to fail. Seems like it might be related to other tests interacting with these. Expanding the testing I'm doing to see if I can track down what specifically is require to get it to fail. I'd say we can move this to WIP

**Comment by NorseGaud** [comment]
@casperklein , I tried `tail -f --pid="$(</var/spool/postfix/pid/master.pid)" /dev/null` in the postfix-wrapper and it causes:

```
❯ make tests
 ✗ File tests.bats
   (from function `repeat_until_success_or_timeout' in file test/test_helper/common.bash, line 37,
    from function `repeat_in_container_until_success_or_timeout' in file test/test_helper/common.bash, line 70,
    from function `wait_for_changes_to_be_detected_in_container' in file test/test_helper/common.bash, line 184,
    from function `setup_file' in test file test/tests.bats, line 49)
     `wait_for_changes_to_be_detected_in_container mail' failed
   596b509db460b28a43e8564cd30437a2b9f1c2ad50f70ccf145abd4c771d5b1f
   [ TASKLOG ]  mail.my-domain.com is up and running
   Timed out on command: docker exec mail bash -c source /usr/local/bin/helper-functions.sh; cmp --silent -- <(_monitored_files_checksums) "${CHKSUM_FILE}" >/dev/null
   mail
   bats warning: Executed 1 instead of expected 16 tests

16 tests, 1 failure, 15 not run
```

I'm going to revert it so that I can continue testing. Maybe we could test this in a separate PR later on.

**Comment by casperklein** [comment]
Ok.. It was worth a try.

**Comment by NorseGaud** [comment]
Ok I think I've narrowed it down: The "setup.sh" tests are causing a situation where port 25 is not active fast enough and therefore the tests trying to use it fail. This is directly related to the changes in https://github.com/docker-mailserver/docker-mailserver/pull/2104 I believe. Continuing to look for a good solution...

**Comment by NorseGaud** [comment]
I can finally reproduce it on macOS by adding to the postfix-wrapper:

```
sleep 10

service postfix start
```

**Comment by NorseGaud** [comment]
The more I think about this, the more I think it's specific to our tests and NOT a real problem. What real-world situation would try to `nc 0.0.0.0 25 < /tmp/docker-mailserver-test/auth/added-smtp-auth-spoofed.txt` and not check if it's available first?

**Comment by casperklein** [comment]
In real world, you would wait for the correct server replies and not send "all" at a time.

**Comment by NorseGaud** [comment]
Adding `wait_for_smtp_port_in_container mail` to the test seemed to work locally but still fails in the CI... Back to trying to figure out what is causing this... :(

**Comment by casperklein** [comment]
Instead of "just" checking if the port is open, you could also check for a basic response:

```bash
until [[ $(echo QUIT | nc localhost 25) == *"221 2.0.0 Bye"* ]]; do
    echo -n .
    sleep 1
done
```

File: test/test_helper/common.bash
Comment (casperklein) [inline test/test_helper/common.bash:90]: `until [[ $(docker exec "${1}" timeout 10 /bin/sh -c "echo QUIT | nc localhost 25") == *"221 2.0.0 Bye"* ]]; do`

If, for whatever reason, nc "blocks" (port is open, but no response from the service), abort after 10 seconds. Otherwise, netcat will just "hang".

Comment (NorseGaud) [reply test/test_helper/common.bash:90]: Yep, I wanted to be sure it worked first. I'll add the count of retries and fail now

Comment (casperklein) [reply test/test_helper/common.bash:91]: I think you missed the "timeout" command in my example?

Comment (NorseGaud) [reply test/test_helper/common.bash]: Oh yes 👍🏼

**Comment by NorseGaud** [comment]
Hmm... Reverting bats seems to break it again...

File: .github/workflows/test_merge_requests.yml
Comment (georglauterbach) [inline .github/workflows/test_merge_requests.yml:64]: Interesting change. On my phone, these lines look identical :D I guess a tab vs. spaces thing. Looks funny though:)

**Review by georglauterbach — DISMISSED** [review]
LGTM 👍🏼

Very nice additions!
I think this PRs paves the way for bringing some of the logic of `setup.sh` in the container itself to make `setup.sh` independent of the mail server version again. I will have a look after this PR is merged :D

File: .github/workflows/test_merge_requests.yml
Comment (NorseGaud) [reply .github/workflows/test_merge_requests.yml:64]: Oh weird... I can bring in the same file from master and it doesn't show any changes to commit... Unsure how this happened 🤔

Comment (NorseGaud) [reply .github/workflows/test_merge_requests.yml:64]: > LGTM 👍🏼
> 
> Very nice additions!
> I think this PRs paves the way for bringing some of the logic of `setup.sh` in the container itself to make `setup.sh` independent of the mail server version again. I will have a look after this PR is merged :D

Thanks! This can be introduced in the next minor release. How do we handle preparing things for certain versions?

**Comment by NorseGaud** [comment]
@georglauterbach 

> LGTM 👍🏼
> 
> Very nice additions!
> I think this PRs paves the way for bringing some of the logic of `setup.sh` in the container itself to make `setup.sh` independent of the mail server version again. I will have a look after this PR is merged :D

Thanks! Newbie maintainer question: This can be introduced in the next minor release. How do we handle preparing things for certain versions?

**Review by casperklein — APPROVED** [review]

**Comment by casperklein** [comment]
The tests seem not to finish 😕

**Comment by NorseGaud** [comment]
> The tests seem not to finish 😕

Sorry, can you clarify what you're seeing?

**Comment by casperklein** [comment]
The build-and-test action took incredible long (1h 28m 58s), but have finished successful 👍

**Comment by NorseGaud** [comment]
> The build-and-test action took incredible long (1h 28m 58s), but have finished successful 👍

Got it -- I see now.

First that that's happened. I'll add some debug output to the tests and see if we can reproduce it.

**Comment by NorseGaud** [comment]
Hmm, it didn't happen before the commit you made. I wonder how much of this is the PR and not just a random/rare situation.

**Comment by NorseGaud** [comment]
Can't reproduce it... Weird...

**Review by georglauterbach — APPROVED** [review]

**Review by casperklein — APPROVED** [review]

**Comment by casperklein** [comment]
Glitch in the matrix or our CI was just busy with other stuff 😉

**Comment by NorseGaud** [comment]
> Glitch in the matrix or our CI was just busy with other stuff 😉

matrix

**Comment by georglauterbach** [comment]
@NorseGaud can this be rebased and merged?

**Comment by NorseGaud** [comment]
> @NorseGaud can this be rebased and merged?

Yep, it's ready. Are we doing 10.2.0?

**Comment by georglauterbach** [comment]
I drafted `v10.2.0`, but the milestone has one or two PRs I'd like to see merged :) I will merge this and test `:edge` until `v10.2.0` is released.

**Pull request merged**
